### ISSUE TITLE
range for floats

### ISSUE THREAD

**Issue by BebeSparkelSparkel** [erikrose/more-itertools#126]
I didn't see a float range function in the library so frange could be helpful. I also used the recommendation from itertools.count to reduce float error.

> When counting with floating point numbers, better accuracy can sometimes be achieved by substituting multiplicative code such as: `(start + step * i for i in count())`.

```python
import itertools, operator

# frange(stop)
# frange(start, stop[, step])
def frange(*args):
  if len(args) == 1:
    start = 0
    stop = args[0]
    step = 1
  elif len(args) == 2:
    start, stop = args
    step = 1
  elif len(args) == 3:
    start, stop, step = args
  else:
    raise TypeError('frange expected at most 1 - 3 arguments, got {}.'.format(len(args)))

  if start < stop and 0 < step:
    compare_with = operator.lt
  elif start > stop and 0 > step:
    compare_with = operator.gt
  else:
    return

  compare_with = operator.lt if start < stop else operator.gt

  for step_count in itertools.count():
    val = start + step * step_count
    if compare_with(val, stop):
      yield val
    else:
      break


import unittest

class Testfrange(unittest.TestCase):
  def test_frange(self):
    self.assertEqual(
        tuple(frange(5)),
        tuple( range(5))
      )

    self.assertEqual(
        tuple(frange(-5)),
        tuple( range(-5))
      )

    self.assertEqual(
        tuple(frange(2, 5)),
        tuple( range(2, 5))
      )

    self.assertEqual(
        tuple(frange(2, 10, 2)),
        tuple( range(2, 10, 2))
      )

    self.assertEqual(
        tuple(frange(2, -5)),
        tuple( range(2, -5))
      )

    self.assertEqual(
        tuple(frange(2, -10, -2)),
        tuple( range(2, -10, -2))
      )

    self.assertEqual(
        tuple(frange(2, 10, -2)),
        tuple( range(2, 10, -2))
      )

    self.assertEqual(
        tuple(frange(2, -10, 2)),
        tuple( range(2, -10, 2))
      )

    self.assertEqual(
        tuple(frange(2.5, 4, 0.5)),
        (2.5, 3, 3.5)
      )

    self.assertEqual(
        tuple(frange(2.5, 4.1, 0.5)),
        (2.5, 3, 3.5, 4)
      )

    with self.assertRaises(TypeError):
      tuple(frange())

    with self.assertRaises(TypeError):
      tuple(frange(1, 2, 3, 4))


if __name__ == '__main__':
  unittest.main()
```

**Comment by bbayles** [comment]
Thanks for the suggestion. I'm surprised this hasn't been proposed before; this is a staple on Stack Overflow and ActiveState.

It feels like we should be able to use `takewhile()` with `count()` here, yes? I'd like to see that implemented in a PR for inclusion.

**Comment by BebeSparkelSparkel** [comment]
I agree with the takewhile but the docs for count suggest that (start + step * i for i in count()) be used instead.

**Comment by bbayles** [comment]
I meant use `takewhile()` with that construct. Something like this:

```python
from itertools import count, takewhile

def frange(*args):
    if len(args) == 1:
        start = 0
        stop, = args
        step = 1
    elif len(args) == 2:
        start, stop = args
        step = 1
    elif len(args) == 3:
        start, stop, step = args
    else:
        raise TypeError(
            'frange expected at most 3 arguments, got {}.'.format(len(args))
        )

    values = (start + (step * n) for n in count())

    if start < stop and 0 < step:
        return takewhile(lambda x: x < stop, values)
    elif start > stop and 0 > step:
        return takewhile(lambda x: x > stop, values)

    return iter(())
```

I'll make a PR for this; I'd like to write a docstring.

**Comment by bbayles** [comment]
PR #127 has an implementation of this. I've made sure the docstring shows that this applies to more than just `float` types, but also `Decimal` and `Fraction` types.

**Comment by bbayles** [comment]
Merged in PR #127 as `numeric_range()`. This will be in the 3.1.0 release.
